src/Formats/CapnProtoSerializer.cpp

```diff
@@ -0,0 +1,1479 @@
+    {
+        const auto & fields = struct_schema.getFields();
+        for (auto field : fields)
+        {
+            auto field_name = String(field.getProto().getName());
+            if (boost::to_lower_copy(name) == boost::to_lower_copy(field_name))
+                return field;
+        }
+        return std::nullopt;
+    }
+
+    [[noreturn]] void throwCannotConvert(const DataTypePtr & type, const String & name, const capnp::Type & capnp_type)
+    {
+        throw Exception(
+            ErrorCodes::CAPN_PROTO_BAD_CAST,
+            "Cannot convert ClickHouse column \"{}\" with type {} to CapnProto type {}",
+            name,
+            type->getName(),
+            getCapnProtoFullTypeName(capnp_type));
+    }
+
+    struct FieldBuilder
+    {
+        virtual ~FieldBuilder() = default;
+    };
+
+    struct ListBuilder : public FieldBuilder
+    {
+        explicit ListBuilder(capnp::DynamicValue::Builder builder, UInt32 elements_size) : impl(builder.as<capnp::DynamicList>()), nested_builders(elements_size)
+        {
+        }
+
+        capnp::DynamicList::Builder impl;
+        std::vector<std::unique_ptr<FieldBuilder>> nested_builders;
+    };
+
+    struct StructBuilder : public FieldBuilder
+    {
+        explicit StructBuilder(capnp::DynamicStruct::Builder struct_builder, size_t fields_size) : impl(std::move(struct_builder)), field_builders(fields_size)
+        {
+        }
+
+        capnp::DynamicStruct::Builder impl;
+        std::vector<std::unique_ptr<FieldBuilder>> field_builders;
+    };
+
+    template <typename ParentBuilder>
+    std::unique_ptr<StructBuilder> initStructBuilder(ParentBuilder & parent_builder, UInt32 offset_or_index, const capnp::_::StructSize & struct_size, size_t elements, const capnp::StructSchema & schema)
+    {
+        capnp::DynamicStruct::Builder builder_impl;
+        if constexpr (std::is_same_v<ParentBuilder, capnp::DynamicStruct::Builder>)
+            builder_impl = capnp::DynamicStruct::Builder(schema, parent_builder.getBuilderImpl().getPointerField(offset_or_index).initStruct(struct_size));
+        else
+            builder_impl = capnp::DynamicStruct::Builder(schema, parent_builder.getBuilderImpl().getStructElement(offset_or_index));
+        return std::make_unique<StructBuilder>(std::move(builder_impl), elements);
+    }
+
+    class ICapnProtoSerializer
+    {
+    public:
+        /// Write row as struct field.
+        virtual void writeRow(
+            const ColumnPtr & column,
+            std::unique_ptr<FieldBuilder> & builder,
+            capnp::DynamicStruct::Builder & parent_struct_builder,
+            UInt32 slot_offset,
+            size_t row_num) = 0;
+
+        /// Write row as list element.
+        virtual void writeRow(
+            const ColumnPtr & column,
+            std::unique_ptr<FieldBuilder> & builder,
+            capnp::DynamicList::Builder & parent_list_builder,
+            UInt32 array_index,
+            size_t row_num) = 0;
```

Comment:
It's not used in simple types serializers, but it's used in Complex types like CapnProtoNullableSerializer, CapnProtoArraySerializer, CapnProtoMapSerializer and CapnProtoStructureSerializer to initialize Capnp structure/list. For example:
https://github.com/ClickHouse/ClickHouse/blob/f39862c3ff7a8f47991a240e3d071aae0cbc6c31/src/Formats/CapnProtoSerializer.cpp#L1303-L1307

We pass builder as an argument because for Capnp struct we should initialize its builder only once (otherwise the second initialize clear previous values), but we can call `writeRow` for the same struct several times, becacuse we want to support writing nested fields separately like `struct.field1 UInt32, struct.field4 String`.